### PR TITLE
Exercise: Run-length encoding (and decoding)

### PR DESCRIPTION
**Pull request by kotp**
~~This is currently draft/unfinished/work-in-progress.~~

Ready for review.

@BNAndras

### CODE REVIEW

**Comment by kotp** [comment]
I will work on replicating the tests (as presented via Python) in a few hours.

**Comment by BNAndras** [comment]
@kotp. did you want me to take a look this weekend? I suspect the problem might be with `substitute`.

**Comment by kotp** [comment]
> @kotp. did you want me to take a look? I'm free this weekend. I suspect the problem might be with `substitute`.

Yeah, holidays around here and just have not had time.  I would appreciate it!

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  AssertEqual g:expected, g:actual
+
+Execute (encode text with no single characters):
+  let g:actual = Encode('AABBBCCCC')
+  let g: expected = '2A3B4C'
+  AssertEqual g:expected, g:actual
+
+Execute (encode single characters with repeated characters):
+  let g:actual = Encode('WWWWWWWWWWWWBWWWWWWWWWWWWBBBWWWWWWWWWWWWWWWWWWWWWWWWB')
+  let g: expected = '12WB12W3B24WB'
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:48]:
```suggestion
  let g:expected = '12WB12W3B24WB'
```

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  AssertEqual g:expected, g:actual
+
+Execute (encode single characters only are encoded without count):
+  let g:actual = Encode('XYZ')
+  let g:expected = 'XYZ'
+  AssertEqual g:expected, g:actual
+
+Execute (encode text with no single characters):
+  let g:actual = Encode('AABBBCCCC')
+  let g: expected = '2A3B4C'
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:43]:
```suggestion
  let g:expected = '2A3B4C'
```

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  let g:actual = Encode('  hsqq qww  ')
+  let g:expected = '2 hs2q q2w2 '
+  AssertEqual g:expected, g:actual
+
+Execute (encode lowercase characters):
+  let g:actual = Encode('aabbbcccc')
+  let g:expected = '2a3b4c'
+  AssertEqual g:expected, g:actual
+
+Execute (encde and then decode empty text gives empty text):
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:61]:
```suggestion
Execute (encode and then decode empty text gives empty text):
```

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  let g:expected = 'A'
+  AssertEqual g:expected, g:actual
+
+Execute (encode two letters that repeat):
+  let g:actual = Encode('aa')
+  let g:expected = '2a'
+  AssertEqual g:expected, g:actual
+
+Execute (encode three letters that repeat):
+  let g:encoded = Encode('aaa')
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:17]:
```suggestion
  let g:actual = Encode('aaa')
```

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  let g:expected = '3a'
+  AssertEqual g:expected, g:actual
+
+Execute (encode abaabbbabaaa):
+  let g:encoded = Encode('abaabbbabaaa')
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:22]:
```suggestion
  let g:actual = Encode('abaabbbabaaa')
```

File: exercises/practice/run-length-encoding/run_length_encoding.vader [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  let g:expected = 'ab2a3bab3a'
+  AssertEqual g:expected, g:actual
+
+Execute (encode a very long text):
+  let g:encoded = Encode('aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz')
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vader:27]:
```suggestion
  let g:actual = Encode('aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz')
```

**Comment by BNAndras** [comment]
I sent you a patch over Discord that updates the test suite and example file. The stub needs to be updated so function names match the test suite.

File: exercises/practice/run-length-encoding/run_length_encoding.vim [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+"" "AABCCCDEEEE"  ->  "2AB3CD4E"  ->  "AABCCCDEEEE"
+""
+"" For simplicity, you can assume that the unencoded string will only contain
+"" the letters A through Z. (either upper or lower case) and whitespace.
+
+"" This means that the encoded data will not contain any numbers, so that any
+"" numbers contained within data to be decoded will always represent the count
+"" for the character that follows it.
+""
+function! RunLengthDecoding(name) abort
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vim:27]:
```suggestion
function! Decode(string) abort
```

File: exercises/practice/run-length-encoding/run_length_encoding.vim [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  " your code goes here
+endfunction
+
+function! RunLengthEncoding(name) abort
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vim:31]:
```suggestion
function! Encode(string) abort
```

File: exercises/practice/run-length-encoding/run_length_encoding.vim [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  " your code goes here
+endfunction
```
Comment (BNAndras) [inline exercises/practice/run-length-encoding/run_length_encoding.vim:33]:
```suggestion
endfunction

```

**Review by BNAndras — APPROVED** [review]

**Comment by kotp** [comment]
@exercism/staff Looks like we may want to have @BNAndras added as a maintainer for this repository.

**Comment by ErikSchierboom** [comment]
@kotp Done!

**Review by BNAndras — APPROVED** [review]

**Pull request merged**